File: extra_credit/ruby/day1.rb
def main
  lhs = []
  rhs = []
  File.foreach('../../day-01/input1_bigger.txt') do |line|
    parts = line.split
    lhs << parts[0].to_i
    rhs << parts[1].to_i
  end
  
  result = lhs.sort.zip(rhs.sort).sum { |v| (v[0] - v[1]).abs }
  puts result
end

def main_chat
  # Get file size to pre-allocate arrays efficiently
  file_size = File.size('../../day-01/input1_bigger.txt')
  estimated_lines = file_size / 16  # Rough estimate of average line length
  
  left = Array.new(estimated_lines)
  right = Array.new(estimated_lines)
  i = 0
  
  # Single pass through file with minimal allocations
  File.open('../../day-01/input1_bigger.txt') do |f|
    f.each_line do |line|
      # Faster than split for exactly 2 numbers
      first_space = line.index(' ')
      left[i] = line[0...first_space].to_i
      right[i] = line[first_space + 3..-1].to_i
      i += 1
    end
  end

  # Trim arrays to actual size
  left.compact!
  right.compact!
  
  # Calculate result
  result = left.sort!.zip(right.sort!).sum { |a, b| (a - b).abs }
  puts result
end

main_chat if $PROGRAM_NAME == __FILE__

File: extra_credit/python/day1.py
def numbers():

    with open("../../day-01/input1_bigger.txt", "r") as fh:
        for line in fh:
            parts = line.split()
            yield (int(parts[0]), int(parts[1]))


def main():
    lhs = list()
    rhs = list()
    with open("../../day-01/input1_bigger.txt", "r") as fh:
        for line in fh:
            parts = line.split()
            lhs.append(int(parts[0]))
            rhs.append(int(parts[1]))

    lhs.sort()
    rhs.sort()
    diff = sum(abs(l - r) for l, r in zip(lhs, rhs))
    print(diff)


def main_ja():
    group_1 = list()
    group_2 = list()
    with open("../../day-01/input1_bigger.txt", "r") as csv_file:
        # csv_reader = csv.reader(csv_file, delimiter=" ")
        for row in csv_file:
            parts = row.split()
            group_1.append(int(parts[0]))
            group_2.append(int(parts[1]))

    group_1.sort()
    group_2.sort()
    difference = sum([abs(g1 - g2) for g1, g2 in zip(group_1, group_2)])
    print(difference)


if __name__ == "__main__":
    main()
